### PR TITLE
Split the VC into crates making it more modular

### PR DESCRIPTION
**Pull request by AgeManning**
# Modularize the VC

The VC currently is built mostly as a single module. It self-references itself making it hard to modularize its components. 

This PR splits the VC into multiple sub-crates that can be imported independently making it useful to external projects, but also making the type dependencies cleaner and the code more logically grouped.

### CODE REVIEW

**Comment by AgeManning** [comment]
I modified the general config, to include some sub-configs (It was required in order to split up dependencies and avoid circular ones). 

The only thing I can really foresee potentially breaking is serializing and de-serializing the validator config. I'm flattening the new inner configs, so I think it should behave the same, but I've not tested it.

**Comment by michaelsproul** [comment]
I reverted the `cargo update` changes so we can get CI to pass on this branch. There's a regression caused by one of our deps:

- https://github.com/sigp/lighthouse/pull/6554

File: validator_client/beacon_node_fallback/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+serde = { workspace = true }
+slog = { workspace = true }
+slot_clock = { workspace = true }
+strum = { workspace = true }
+tokio = { workspace = true }
+types = { workspace = true }
+validator_metrics = { workspace = true }
+
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/beacon_node_fallback/Cargo.toml]: Can just delete this

```suggestion
```

File: validator_client/http_api/src/tests.rs [hunk 2]
```diff
@@ -46,43 +46,39 @@ struct ApiTester {
...
     }
 
-    pub async fn new_with_config(mut config: Config) -> Self {
+    pub async fn new_with_config(config: ValidatorStoreConfig) -> Self {
         let log = test_logger();
 
-        let validator_dir = tempdir().unwrap();
-        let secrets_dir = tempdir().unwrap();
+        let validator_dir: PathBuf = tempdir().unwrap().path().into();
+        let secrets_dir: PathBuf = tempdir().unwrap().path().into();
```
Comment (michaelsproul) [inline validator_client/http_api/src/tests.rs]: I don't think this is a good idea, as the tempdir will get dropped and deleted?

I'm surprised the tests are passing currently

File: validator_client/http_metrics/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+slog = { workspace = true }
+warp_utils = { workspace = true }
+warp = { workspace = true }
+lighthouse_version = { workspace = true }
+validator_services = { workspace = true }
+validator_store = { workspace = true }
+validator_metrics = { workspace = true }
+types = { workspace = true }
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/http_metrics/Cargo.toml]:
```suggestion
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Reviewed half, will come back for the rest tomorrow

File: validator_client/http_metrics/src/lib.rs [hunk 4]
```diff
@@ -156,3 +158,57 @@ pub fn serve<E: EthSpec>(
...
+
+    // It's important to ensure these metrics are explicitly enabled in the case that users aren't
+    // using glibc and this function causes panics.
+    if ctx.config.allocator_metrics_enabled {
+        scrape_allocator_metrics();
+    }
+
+    warp_utils::metrics::scrape_health_metrics();
+
+    encoder.encode(&metrics::gather(), &mut buffer).unwrap();
```
Comment (michaelsproul) [inline validator_client/http_metrics/src/lib.rs]:
```suggestion
    encoder.encode(&metrics::gather(), &mut buffer).map_err(|e| format!("{:e?}"))?;
```

File: validator_client/initialized_validators/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+serde = { workspace = true }
+serde_json = { workspace = true }
+bls = { workspace = true }
+tokio = { workspace = true }
+bincode = { workspace = true }
+filesystem = { workspace = true }
+validator_metrics = { workspace = true }
+
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/initialized_validators/Cargo.toml]:
```suggestion
```

File: validator_client/signing_method/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+parking_lot = { workspace = true }
+reqwest = { workspace = true }
+task_executor = { workspace = true }
+types = { workspace = true }
+url = { workspace = true }
+validator_metrics = { workspace = true }
+serde = { workspace = true }
+ethereum_serde_utils = { workspace = true }
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/signing_method/Cargo.toml]:
```suggestion
```

File: validator_client/src/check_synced.rs
Comment (michaelsproul) [inline validator_client/src/check_synced.rs:1]: Nice catch deleting this unused file!

File: validator_client/validator_metrics/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+authors = ["Sigma Prime <contact@sigmaprime.io>"]
+
+[lib]
+name = "validator_metrics"
+path = "src/lib.rs"
+
+[dependencies]
+metrics = { workspace = true }
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/validator_metrics/Cargo.toml]:
```suggestion
```

File: validator_client/validator_services/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+parking_lot = { workspace = true }
+safe_arith = { workspace = true }
+slog =  { workspace = true }
+slot_clock =  { workspace = true }
+tokio =  { workspace = true }
+types =  { workspace = true }
+tree_hash =  { workspace = true }
+bls =  { workspace = true }
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/validator_services/Cargo.toml]:
```suggestion
```

File: validator_client/validator_store/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+serde = { workspace = true }
+signing_method = { workspace = true }
+slashing_protection = { workspace = true }
+slog = { workspace = true }
+slot_clock = { workspace = true }
+task_executor = { workspace = true }
+types = { workspace = true }
+validator_metrics = { workspace = true }
+
+[dev-dependencies]
```
Comment (michaelsproul) [inline validator_client/validator_store/Cargo.toml]:
```suggestion
```

File: validator_client/validator_store/src/lib.rs [hunk 10]
```diff
@@ -48,6 +42,30 @@ impl From<SigningError> for Error {
     }
 }
 
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct Config {
+    /// Fallback fallback address.
```
Comment (michaelsproul) [inline validator_client/validator_store/src/lib.rs]:
```suggestion
    /// Fallback fee recipient address.
```

File: validator_client/validator_store/src/lib.rs [hunk 11]
```diff
@@ -77,6 +95,12 @@ pub struct ValidatorStore<T, E: EthSpec> {
     _phantom: PhantomData<E>,
 }
 
+impl<T: SlotClock + 'static, E: EthSpec> DoppelGangerValidatorStore for ValidatorStore<T, E> {
```
Comment (michaelsproul) [inline validator_client/validator_store/src/lib.rs]: I think we should rename this trait. Doppelganger with a capital G looks really weird

```suggestion
impl<T: SlotClock + 'static, E: EthSpec> DoppelgangerValidatorStore for ValidatorStore<T, E> {
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]
Looks good on the whole. This was less daunting than I expected.

I'll push some small tweaks and then we can merge.

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
@mergify queue

**Pull request merged**